### PR TITLE
[IMP] account: allows multiple payment acquirers on a journal.

### PR DESCRIPTION
**Pull request by vin-odoo**
Users may want to be able to have transactions coming from multiple
payment acquirers to be registered in the same journal.

This will allows that.

Task id #2414749

### CODE REVIEW

File: addons/account/models/account_payment.py
Comment (william-andre) [inline addons/account/models/account_payment.py:40]: `required=True`

File: addons/account/models/account_payment.py
Comment (william-andre) [inline addons/account/models/account_payment.py:20]: Like discussed, define `name_get` to compute the `display_name` instead. (and probably `name_search` too)

File: addons/payment/models/payment_acquirer.py
Comment (william-andre) [inline addons/payment/models/payment_acquirer.py:187]: This should be a computed (non stored?) field
Add a constraint that checks the `journal_id` has a method for this acquirer too

File: addons/payment/models/account_journal.py
Comment (william-andre) [inline addons/payment/models/account_journal.py:11]: I don't get any of this

**Review by william-andre — COMMENTED** [review]
The biggest thing right now is that `journal_id` should be required on `account.payment.type`

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:797]: You don't need to assign specificaly your record to a one2many. This is made automatically by the orm.

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:786]: Not sure this method is really helpful. Making ```
bank_journal.outbound_payment_method_id = [Command.create({'method_type': <my_method>.id})]
``` is cleaner and less spaghetti in `_enable_check_printing_on_bank_journals` imo

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:28]: move it to a dedicated file plz
Same for views

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:31]: I'm not sure this is really necessary.

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:33]: `default=10`

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:35]: BY default, there is no `electronic` right? This should be added by the `payment` module I guess

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:32]: `active` is kind of protected field managing the `archive` smartbutton. Is it really what you want? To discuss

File: addons/account/views/account_journal_views.xml
Comment (smetl) [inline addons/account/views/account_journal_views.xml:87]: I think `id="inbound_payment_settings"` is better

File: addons/account/views/account_journal_views.xml
Comment (smetl) [inline addons/account/views/account_journal_views.xml:101]: same here

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:59]: not sure about all these related fields. Btw you should split the fields like `Business fields` / `Display purpose fields`

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:494]: `payment_debit_account_id` and `payment_credit_account_id` no longer exist on `account.journal`

Comment (smetl) [reply addons/account/models/account_payment.py:494]: Why not `pay.outstanding_account_id = pay.payment_method_id.account_id or pay.company_id.<default_account_in_the_company_but_I_dont_remember_its_name>`

**Review by smetl — CHANGES_REQUESTED** [review]
Review of the `account` part

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:237]: you should assign `inbound_payment_method_line_ids` in any case:
```
pay_method_line_ids_commands = [Command.clear()]
if journal.type in ('bank', 'cash'):
    for default_method in journal._default_inbound_payment_methods():
        pay_method_line_ids_commands.append(Command.create({'payment_method_id': default_method.id})
journal.inbound_payment_method_line_ids = pay_method_line_ids_commands
```

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:583]: These fields seem useless... I didn't find any usage of them. To be removed? :p

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:702]: Missing `self.ensure_one()`
You should not fallback on the company if not needed.
```
need_fallback = False
for line in self.inbound_payment_method_line_ids:
     account_ids.add(line.payment_account_id.id or self.company_id.account_journal_payment_debit_account_id.id)        
return self.env['account.account'].browse(account_ids)
```
Note: If `account_journal_payment_debit_account_id` is not required, don't add `False` into `account_ids`

Comment (smetl) [reply addons/account/models/account_journal.py:702]: Note: Are these methods still necessary?

File: addons/account/models/account_payment.py
Comment (smetl) [inline addons/account/models/account_payment.py:424]: I don't know if we should keep this fallback on `default_account_id`. This could be very dangerous for the end user regarding the bank statements management

File: addons/account/models/account_payment_method.py
Comment (smetl) [inline addons/account/models/account_payment_method.py:51]: You don't need to specify ´readonly=True´ for computed field since this is the default value

File: addons/account/tests/common.py
Comment (smetl) [inline addons/account/tests/common.py:186]: Thanks to the default, this setup shouldn't be necessary, right?

File: addons/account/views/account_journal_views.xml
Comment (smetl) [inline addons/account/views/account_journal_views.xml:95]: can you inline in columns plz? Too many kwargs

File: addons/account/views/account_payment_method_views.xml
Comment (smetl) [inline addons/account/views/account_payment_method_views.xml:33]: I don't understand the purpose of such views.
Btw, plz add `groups="account.group_account_readonly"`

File: addons/account/wizard/account_payment_register_views.xml
Comment (smetl) [inline addons/account/wizard/account_payment_register_views.xml:33]: `'not_open': True` ?

File: addons/account_check_printing/models/account_journal.py
Comment (smetl) [inline addons/account_check_printing/models/account_journal.py:18]: See my comment above:
```
res = super()._default_outbound_payment_methods()
if self.type == 'bank':
    res |= self.env.ref('account_check_printing.account_payment_method_check')
return res
```

File: addons/payment/models/payment_acquirer.py
Comment (smetl) [inline addons/payment/models/payment_acquirer.py:199]: typo.

File: addons/payment/models/payment_acquirer.py
Comment (smetl) [inline addons/payment/models/payment_acquirer.py:138]: you don't need this field

File: addons/payment/models/payment_acquirer.py
Comment (smetl) [inline addons/payment/models/payment_acquirer.py:197]: I think we should avoid this behavior if the acquirer is in `test` or `enabled`. Then, you could rewrite it with:
```
payment_method = self.env['account.payment.method.line'].search([('code', '=', acquirer.provider)])
payment_method.journal_id = acquirer.journal_id
```

File: addons/payment/models/payment_acquirer.py
Comment (smetl) [inline addons/payment/models/payment_acquirer.py:280]: this method could be removed, right?

File: addons/point_of_sale/security/ir.model.access.csv
Comment (smetl) [inline addons/point_of_sale/security/ir.model.access.csv:55]: wtf?

File: addons/test_website_slides_full/tests/test_ui_wslides.py
Comment (smetl) [inline addons/test_website_slides_full/tests/test_ui_wslides.py:162]: wtf?

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:238]: You could use `available_payment_method_ids` there. 
`for pay_method in journal.available_payment_method_ids.filtered(lambda x: x.type == 'inbound')`

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:726]: Let's use `_get_journal_inbound_outstanding_payment_accounts` / `_get_journal_outbound_outstanding_payment_accounts` here

File: addons/account/models/account_payment_method.py
Comment (smetl) [inline addons/account/models/account_payment_method.py:29]: I think we should force anyone to provide method_information for the newly created payment_method

File: addons/account/models/account_payment_method.py
Comment (smetl) [inline addons/account/models/account_payment_method.py:36]: payment_methods are shared accross multiple companies ;)

Comment (smetl) [reply addons/account/models/account_payment_method.py:36]: this should be enough:
```
default_domain = [('type', 'in', ('bank', 'cash'))]
domain = information.get('domain', default_domain)
domains = [domain]
if currency_id:
    domains.add(expression.OR([
        [('currency_id', '=', False), ('company_id.currency_id', '=', currency_id)],
        [('currency_id', '=', currency_id)],
    )
full_domain = expression.AND(domains)
```

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:200]: `_get_payment_method_information` ?

File: addons/payment/wizards/account_payment_register.py
Comment (smetl) [inline addons/payment/wizards/account_payment_register.py:44]: same comment here :)

File: addons/account_check_printing/models/account_payment_method.py
Comment (smetl) [inline addons/account_check_printing/models/account_payment_method.py:12]: Fine tuning: Since python 3.0, it could be rewritten by `res = super()._get_payment_method_information()`

File: addons/account_check_printing/tests/test_print_check.py
Comment (smetl) [inline addons/account_check_printing/tests/test_print_check.py:30]: Since `account_payment_method_check` is enabled by default, do you really need such kind of diff now?

File: addons/sale/tests/test_payment_flow.py
Comment (smetl) [inline addons/sale/tests/test_payment_flow.py:11]: to remove

File: addons/point_of_sale/security/ir.model.access.csv
Comment (smetl) [reply addons/point_of_sale/security/ir.model.access.csv:55]: remove this diff plz

File: odoo/fields.py
Comment (smetl) [inline odoo/fields.py:3240]: :fearful:

**Review by smetl — CHANGES_REQUESTED** [review]

File: addons/account/models/account_journal.py
Comment (smetl) [inline addons/account/models/account_journal.py:3]: merge import plz

**Comment by smetl** [comment]
@robodoo r+

**Review by AntoineVDV — COMMENTED** [review]
Hello, this is a welcome change, and it LGTM overall but I have a few remarks, please find them below :)

**Comment by william-andre** [comment]
@AntoineVDV It is already merged :)

File: addons/payment/models/account_journal.py
Comment (william-andre) [reply addons/payment/models/account_journal.py]: I don't agree with this one, it is already done by the SQL query just above it... Why would you do twice the work?

**Review by AntoineVDV — APPROVED** [review]
Thanks!

File: addons/payment/models/account_journal.py
Comment (AntoineVDV) [reply addons/payment/models/account_journal.py:29]: You're right, I read that part too fast! (I was trying to outrun the mergebot :stuck_out_tongue:)

**Comment by smetl** [comment]
@robodoo r+

File: addons/point_of_sale/security/ir.model.access.csv
Comment (Feyensv) [reply addons/point_of_sale/security/ir.model.access.csv:55]: Why is this diff still here ? :thinking:

Comment (william-andre) [reply addons/point_of_sale/security/ir.model.access.csv:55]: I understand that there is a missing or exceeding LF or CR char, I don't know if there is something about that in the guidelines (I would assume all lines should end with `\r\n`, maybe? I don't remember.)
But even if this fixes the guidelines, it should have been done in another commit...

Comment (william-andre) [reply addons/point_of_sale/security/ir.model.access.csv:55]: A bit too late now since this is a priority for the freeze though...

Comment (Feyensv) [reply addons/point_of_sale/security/ir.model.access.csv:55]: Well, while waiting for a review from upgrade, it could be a good time to remove this...

**Comment by william-andre** [comment]
@robodoo r+ rebase-ff

File: addons/point_of_sale/security/ir.model.access.csv
Comment (KangOl) [reply addons/point_of_sale/security/ir.model.access.csv:55]: This PR has conflicts. Take the opportunity to remove this diff.

**Comment by william-andre** [comment]
@robodoo r+

**Comment by nhomar** [comment]
Which use case cover this? If odoo star do thing becaus users want we are in a dark path :-(

how will we reconcile acquires if we mixup different journal…

**Pull request closed without merging**
